flowzone.yml

```diff
@@ -2365,114 +2317,93 @@ jobs:
-                npm ci
-              else
-                npm i
-              fi
+          if [ ${{ needs.is_npm.outputs.has_npm_lockfile }} == 'true' ]; then
+            npm ci
           else
-              echo "${runner_os} is not supported in package.json"
+            npm i
           fi
 
       - name: Run build
+        if: steps.check_engine.outcome == 'success'
         run: npm run build --if-present
 
       - name: Run tests
-        if: inputs.pseudo_terminal != true
+        if: inputs.pseudo_terminal != true && steps.check_engine.outcome == 'success'
         run: npm test
 
       - name: Run tests (pseudo-tty)
-        if: inputs.pseudo_terminal == true
+        if: inputs.pseudo_terminal == true && steps.check_engine.outcome == 'success'
         shell: script -q -e -c "bash --noprofile --norc -eo pipefail -x {0}" /tmp/test-session
         run: npm test
 
       - name: Run pack
-        if: needs.is_npm.outputs.npm_private != 'true' && steps.node_versions.outputs.max == matrix.node_version
+        if: needs.is_npm.outputs.npm_private != 'true' && steps.check_engine.outcome == 'success'
         run: |
           mkdir ${{ runner.temp }}/npm-pack && npm pack --pack-destination=${{ runner.temp }}/npm-pack
 
           # FIXME: workaround when `npm pack` for npm 6.x dumps tarball into the current directory because it has no `--pack-destination` flag
           [[ "$(npm --version)" =~ ^6\..* ]] && find . -maxdepth 1 -name '*.tgz' -exec mv {} ${{ runner.temp }}/npm-pack \; || true
 
       - name: Upload artifact
-        if: needs.is_npm.outputs.npm_private != 'true' && steps.node_versions.outputs.max == matrix.node_version
+        if: needs.is_npm.outputs.npm_private != 'true' && steps.check_engine.outcome == 'success'
         uses: actions/upload-artifact@0b2256b8c012f0828dc542b3febcab082c67f72b # v4.3.4
         with:
-          name: npm-${{ github.event.pull_request.head.sha }}-${{ matrix.node_version }}
+          name: npm-${{ github.event.pull_request.head.sha }}-${{ strategy.job-index }}
           path: ${{ runner.temp }}/npm-pack/*.tgz
           retention-days: 90
 
       - name: Generate docs (if present)
-        if: needs.is_npm.outputs.npm_docs == 'true'
+        if: needs.is_npm.outputs.npm_docs == 'true' && steps.check_engine.outcome == 'success'
         shell: bash
         run: npm run doc
 
       - name: Compress docs
-        if: needs.is_npm.outputs.npm_docs == 'true' && steps.node_versions.outputs.max == matrix.node_version
+        if: needs.is_npm.outputs.npm_docs == 'true' && steps.check_engine.outcome == 'success'
         run: tar --auto-compress -cvf ${{ runner.temp }}/docs.tar.zst ./docs
 
       - name: Upload artifact
-        if: needs.is_npm.outputs.npm_docs == 'true' && steps.node_versions.outputs.max == matrix.node_version
+        if: needs.is_npm.outputs.npm_docs == 'true' && steps.check_engine.outcome == 'success'
```

Comment:
I wonder if we could skip some of these steps since only one artifact will be used in the end

Reply:
I wanted to, but as all the jobs are running in parallel they have no way of knowing whether their artifact is the one that will be used or not.

This is the downside of performing the engine checks as part of the tests, we don't know ahead of time which matrix jobs are expected to pass and create artifacts. I still think the pros outweigh the cons though, especially since these uploads are done in parallel.